File: dsu.cpp
struct DSU{
	int n, components;
	vi par, sz;
	int root(int x)
	{ return x==par[x]?x:(par[x]=root(par[x]));	}
	bool merge(int x1, int x2)
	{
		x1=root(x1), x2=root(x2);
		if(x1==x2) return false;
		if(sz[x2]>sz[x1]) swap(x1, x2);
		par[x2]=x1; sz[x1]+=sz[x2];
		--components;
		return true;
	}
	DSU(int _n)
	{
		n=components=_n;
		sz.assign(n+1, 1);
		par.resize(n+1);
		iota(all(par), 0);
	}
};

File: rust/data_structures/dsu.rs
struct DSU {
    parent:  Vec<usize>,
    component_size: Vec<usize>,
    num_comps: usize,
}

impl DSU {
    fn root(&mut self, x: usize) -> usize {
        if x == self.parent[x] {
            return x;
        }
        self.parent[x] = self.root(self.parent[x]);
        self.parent[x]
    }

    fn merge(&mut self, c1: usize, c2: usize) -> bool {
        let mut r1 = self.root(c1);
        let mut r2 = self.root(c2);
        if r1 == r2 {
            return false;
        }
        if self.component_size[r2] > self.component_size[r1] {
            swap(&mut r1, &mut r2);
        }
        self.parent[r2] = r1;
        self.component_size[r1] += self.component_size[r2];
        self.num_comps -= 1;
        true
    }

    fn new(n: usize) -> DSU {
        DSU {
            parent: (0..n).map(|i| i).collect(),
            component_size: vec![1; n],
            num_comps: n,
        }
    }
}
